### PR TITLE
Add build target for packaging tsserver as a library

### PR DESCRIPTION
**Pull request by weswigham**
Specifically, adds `jake lssl` (language service server library). The goal was to export the `Session` type which, in turn, has necessitated exporting most of the types in `editorServices.ts` by extension. 

Additionally, I've made some changes to enable this library to function outside node - (`node.d.ts` is no longer compiled into the `lssl` target by any of the files included in it). Specifically, the two features of Node which Session used - `Buffer.byteLength` and `process.hrtime`, I've offloaded to a new `Environment` interface, and created a node implementation of it for `tsserver` - `server\nodeimpl.ts`. I was considering that they may be better provided by the `System` interface, and would like feedback on this - changing the `System` interface seemed like a farther reaching change than was desired.

On top of this, to make `Session` (or trivial subclasses) actually instantiate at runtime in non-node environments, I removed `Session`'s meager dependency on `ts.sys` - which only gets set in node or in windows script host. For the most part, this simply meant calling the host which `Session` was constructed with, but in one instance this meant that the `host` had to be passed along to a static function - `ScriptVersionCache.fromString`. Additionally, the static `CompilerService.defaultFormatCodeOptions` had to be made aware that `ts.sys` may not be set at static initialization time if not running in either node or windows script host.

The remaining changes, b8f6ada and 9a3fead, are less critical to exporting the `Session` type as a library, but do provide quality improvements in the exported interface. 
* b8f6ada allows consumers of the `Session` class to subclass and override `onMessage` and still call the `onMessageParsed` implementation - or just call `onMessageParsed` directly. This is useful in situations where the service is capable of receiving message objects as objects directly, rather than as json strings, thereby allowing consumers to remove (de)serialization overhead.
* 9a3fead adds support into the new `onMessageParsed` function to handle arbitrary messages, rather than throwing immediately if the massive switch statement falls all the way through. (IMO, it may be nicer to register all message handlers this way and remove the switch entirely.) This is incredibly useful in situations when the session communication channel is intended to the the primary communication channel between a client and a server, and allows a server to provide extended functionality (or, in the case of running in a browser worker thread, allows a client to send the server content with which to build a virtual filesystem and external edits to the files therein). This is the least required of the changes herein, so if too much issue is taken with it, it is not necessarily required if the prior commit is still accepted.

### CODE REVIEW

File: src/server/nodeimpl.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+/// <reference path="node.d.ts" />
+module ts.server {
+    export class NodeEnvironment implements Environment {
+        byteLength = Buffer.byteLength;
```
Comment (mhegazy) [inline src/server/nodeimpl.ts]: these are only used in the session object, i do not see the point of the additional abstraction, just pass them as constructor paramters.

File: Jakefile.js [hunk 2]
```diff
@@ -100,11 +100,20 @@ var serverSources = [
     "editorServices.ts",
     "protocol.d.ts",
     "session.ts",
+    "nodeimpl.ts",
```
Comment (mhegazy) [inline Jakefile.js]: i do not think you need the extra file if we pass the values as a constructor parameter.

File: Jakefile.js [hunk 3]
```diff
@@ -369,6 +379,20 @@ compileFile(servicesFile, servicesSources,[builtLocalDirectory, copyright].conca
 var serverFile = path.join(builtLocalDirectory, "tsserver.js");
 compileFile(serverFile, serverSources,[builtLocalDirectory, copyright].concat(serverSources), /*prefixes*/ [copyright], /*useBuiltCompiler*/ true);
 
+var lsslFile = path.join(builtLocalDirectory, "tslssl.js");
+compileFile(
+    lsslFile, 
+    languageServiceLibrarySources, 
+    [builtLocalDirectory, copyright].concat(serverSources).concat(languageServiceLibrarySources),
```
Comment (mhegazy) [inline Jakefile.js]: you do not need `.concat(serverSources)`, do you?

File: Jakefile.js [hunk 4]
```diff
@@ -100,11 +100,20 @@ var serverSources = [
     "editorServices.ts",
     "protocol.d.ts",
     "session.ts",
+    "nodeimpl.ts",
     "server.ts"
 ].map(function (f) {
     return path.join(serverDirectory, f);
 });
 
+var languageServiceLibrarySources = [
```
Comment (mhegazy) [inline Jakefile.js]: i would concat the services sources as well, to get incremental compilation working correctly.

File: src/server/editorServices.ts [hunk 5]
```diff
@@ -1263,6 +1262,7 @@ module ts.server {
         versions: LineIndexSnapshot[] = [];
         minVersion = 0;  // no versions earlier than min version will maintain change history
         private currentVersion = 0;
+        private host: System;
```
Comment (mhegazy) [inline src/server/editorServices.ts]: `ServerHost` instead of `System`

File: src/server/session.ts [hunk 6]
```diff
@@ -108,17 +107,34 @@ module ts.server {
 
     export interface ServerHost extends ts.System {
     }
+    
+    export interface Environment {
+        byteLength: (buf: string, encoding?: string) => number;
+        hrtime: (start?: number[]) => number[]; //array of seconds, nanoseconds
+    }
+    
+    export interface Message {
```
Comment (mhegazy) [inline src/server/session.ts]: this is already in `ts.server.protocol`

File: src/server/session.ts [hunk 7]
```diff
@@ -108,17 +107,34 @@ module ts.server {
...
+        byteLength: (buf: string, encoding?: string) => number;
+        hrtime: (start?: number[]) => number[]; //array of seconds, nanoseconds
+    }
+    
+    export interface Message {
+        type: string,
+        seq: number
+    }
+    
+    export interface Event extends Message {
```
Comment (mhegazy) [inline src/server/session.ts]: same here, why do we need to duplicate this?

File: src/server/session.ts [hunk 8]
```diff
@@ -834,153 +850,178 @@ module ts.server {
 
         exit() {
         }
+        
+        private handlers : Map<ProtocolHandler> = {};
```
Comment (mhegazy) [inline src/server/session.ts]: please split this in a different change. i would like to discuss the design first.

File: src/server/session.ts [hunk 9]
```diff
@@ -834,153 +850,178 @@ module ts.server {
...
+        
+        private handlers : Map<ProtocolHandler> = {};
+        addProtocolHandler(command: string, handler: ProtocolHandler) {
+            this.handlers[command] = handler;
+        }
+        removeProtocolHandler(command: string, handler: ProtocolHandler) {
+            delete this.handlers[command];
+        }
+        
+        onMessageParsed(request: protocol.Request) : boolean {
```
Comment (mhegazy) [inline src/server/session.ts]: i am not sure i see the value of splitting `onMessage` into two methods.

Comment (mhegazy) [inline src/server/session.ts]: oh i see.. i would call it executeCommand maybe.. the "on" part is misleading.

**Comment by mhegazy** [comment]
@weswigham can you split this into two, pre https://github.com/Microsoft/TypeScript/commit/b8f6ada5d83113a901db54d0b65bd9e8f9edd73c and https://github.com/Microsoft/TypeScript/commit/9a3fead30d36511ba9e74a75f5688f5b02d01406, i would want to discuss them first.
I do not think we need the extra layer of abstraction for nodeEnviroment.

File: Jakefile.js (hunk 3 above)
Comment (weswigham) [inline Jakefile.js]: Aye, good catch. Didn't change that since removing the extra section up above.

File: Jakefile.js (hunk 4 above)
Comment (weswigham) [inline Jakefile.js]: Alright.

File: src/server/session.ts (hunk 6 above)
Comment (weswigham) [inline src/server/session.ts]: It is? Alright, I'll use it from there. In this file it was being stolen from node.d.ts's debug message format for some reason.

File: src/server/session.ts (hunk 7 above)
Comment (weswigham) [inline src/server/session.ts]: See above.

File: src/server/session.ts (hunk 8 above)
Comment (weswigham) [inline src/server/session.ts]: Can and will do.

File: src/server/session.ts (hunk 9 above)
Comment (weswigham) [inline src/server/session.ts]: Hm. Good idea. I'll change the name.

File: src/server/nodeimpl.ts (hunk 1 above)
Comment (weswigham) [inline src/server/nodeimpl.ts]: That's pretty much what this is doing, it was just bundling the two parameters into one argument. I'll split it out into multiple arguments.

**Comment by weswigham** [comment]
Done - see #3542 for the library packaging part.

**Pull request closed without merging**